src/PerformanceSensitiveAnalyzers/CSharp/CodeFixes/AvoidAllocationWithArrayEmptyCodeFix.cs

```diff
@@ -0,0 +1,283 @@
+        {
+            return IsTypeReadonlySequence(semanticModel, propertyDeclaration.Type);
+        }
+
+        private static bool IsReturnTypeReadonlySequence(SemanticModel semanticModel, MethodDeclarationSyntax methodDeclarationSyntax)
+        {
+            var typeSyntax = methodDeclarationSyntax.ReturnType;
+            return IsTypeReadonlySequence(semanticModel, typeSyntax);
+        }
+
+        private static bool IsExpectedParameterReadonlySequence(SyntaxNode node, SemanticModel semanticModel)
+        {
+            if (node is ArgumentSyntax argument && node.Parent is ArgumentListSyntax argumentList)
+            {
+                var argumentIndex = argumentList.Arguments.IndexOf(argument);
+                if (semanticModel.GetSymbolInfo(argumentList.Parent).Symbol is IMethodSymbol methodSymbol)
+                {
+                    if (methodSymbol.Parameters.Length > argumentIndex)
+                    {
+                        var parameterType = methodSymbol.Parameters[argumentIndex].Type;
+                        if (IsTypeReadonlySequence(semanticModel, parameterType))
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool IsTypeReadonlySequence(SemanticModel semanticModel, TypeSyntax typeSyntax)
+        {
+            var returnType = ModelExtensions.GetTypeInfo(semanticModel, typeSyntax).Type;
+            return IsTypeReadonlySequence(semanticModel, returnType);
+        }
+
+        private static bool IsTypeReadonlySequence(SemanticModel semanticModel, ITypeSymbol type)
+        {
+            if (type.Kind == SymbolKind.ArrayType)
+            {
+                return true;
+            }
+
+            if (type is INamedTypeSymbol namedType && namedType.IsGenericType)
+            {
+                foreach (var readonlySequence in GetReadonlySequenceTypes(semanticModel))
+                {
+                    if (namedType.ConstructedFrom.Equals(readonlySequence))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static readonly ImmutableArray<string> _readonlySequenceTypeNames = ImmutableArray.Create(
+            WellKnownTypeNames.SystemCollectionsGenericIEnumerable1,
+            WellKnownTypeNames.SystemCollectionsGenericIReadOnlyList1,
+            WellKnownTypeNames.SystemCollectionsGenericIReadOnlyCollection1);
+
+        private static IEnumerable<INamedTypeSymbol?> GetReadonlySequenceTypes(SemanticModel semanticModel)
+        {
+            var provider = WellKnownTypeProvider.GetOrCreate(semanticModel.Compilation);
+            return _readonlySequenceTypeNames.Select(name => provider.GetOrCreateTypeByMetadataName(name));
```

Comment:
@mavasani Prefers to keep the `WellKnownTypeProvider` hidden if not necessary.

```suggestion
            return _readonlySequenceTypeNames.Select(name => semanticModel.Compilation.GetOrCreateTypeByMetadataName(name));
```